### PR TITLE
[TypeScript] Support default exports in TSExportAssignment.

### PR DESCRIPTION
**Pull request by joaovieira**
Allows TypeScript export assignments of default identifiers and/or mixed merged declarations. E.g.:

```ts
declare const foobar: number;
export = foobar;
```

```ts
declare function foobar(): void;
declare namespace foobar {
  type MyType = string
}
export = foobar;
```

```ts
import { foobar } from './module';
export = foobar;
```

```ts
// No longer reports missing default export
import foobar, { MyType } from './foobar';
```

Fixes: #1527

### CODE REVIEW

File: src/ExportMap.js [hunk 1]
```diff
@@ -514,30 +514,50 @@ ExportMap.parse = function (path, content, context) {
...
+      const exportedDecls = ast.body.filter(({ type, id, declarations }) => 
+        declTypes.includes(type) && 
+        (id && id.name === exportedName || declarations.find(d => d.id.name === exportedName))
       )
-      moduleDecls.forEach((moduleDecl) => {
-        if (moduleDecl && moduleDecl.body && moduleDecl.body.body) {
-          moduleDecl.body.body.forEach((moduleBlockNode) => {
+      if (exportedDecls.length === 0) {
+        // Export is not referencing any local declaration, must be re-exporting
+        m.namespace.set('default', captureDoc(source, docStyleParsers, n))
```
Comment (joaovieira) [inline src/ExportMap.js]: I'm not sure about `n` here. Would it be worth looking through the `m.imports` to see if any matches the exported name? We don't do that on [regular exports](https://github.com/benmosher/eslint-plugin-import/pull/1528/files#diff-e991b36c6628220049668b5e2d2c43d8R501). E.g.:

```
import { test } from './test';
export { foo }; // all good...
```

**Comment by JounQin** [comment]
It should read about `esModuleInterop` option or add an equal option for this plugin instead of hard coding the logic.

**Comment by ljharb** [comment]
@JounQin can you elaborate?

TS's module system is broken without synthetic imports and esModuleInterop enabled; I'm completely comfortable telling people they need to enable those if they want stuff to work properly.

**Comment by Maxim-Mazurok** [comment]
What a great PR! I was searching for a solution all day, finally gave up and started to dig into the sources. Strangely enough, this PR didn't come up while I was searching for similar issues.
So, I started coding basically the same thing: #1689

@JounQin I agree with you, in up PR I'm reading from `tsconfig.json`. But, there might be multiple tsconfigs and they might include/exclude specific files. So, the best approach would be to first, figure out which tsconfig is covering file where the import is made and then check if that config has `esModuleInterop` enabled.
Here's an [example](https://github.com/alexgorbatchev/eslint-import-resolver-typescript/blob/585509e95f93adf8b7ef5839029c19c55edbe76e/src/index.ts#L148-L171) of code from `eslint-import-resolver-typescript` that does kind of similar thing: 
```js
mappers = configPaths!
  // turn glob patterns into paths
  .reduce<string[]>(
    (paths, path) => paths.concat(isGlob(path) ? globSync(path) : path),
    [],
  )
  .map(loadConfig)
  .filter(isConfigLoaderSuccessResult)
  .map(configLoaderResult => {
    const matchPath = createMatchPath(
      configLoaderResult.absoluteBaseUrl,
      configLoaderResult.paths,
    )

    return (source: string, file: string) => {
      // exclude files that are not part of the config base url
      if (!file.includes(configLoaderResult.absoluteBaseUrl)) {
        return undefined
      }

      // look for files based on setup tsconfig "paths"
      return matchPath(source, undefined, undefined, extensions)
    }
  })
```

That being said, we should either do the right thing and figure out how to handle complex cases. Or, we can do the simple thing and only cover simple case (only root tsconfig) or introduce new leg in `parserOptions` that will also act globally but will be more explicit than assuming that everybody has only one tsconfig.

Also, I can't disagree with @ljharb about assuming that everyone uses `esModuleInterop` nowadays. It seems to be a pretty popular flag to use. And in this case, I would be very happy to see this PR merged, thanks!

File: src/ExportMap.js [hunk 2]
```diff
@@ -530,30 +530,50 @@ ExportMap.parse = function (path, content, context) {
...
+        'TSDeclareFunction',
+        'TSEnumDeclaration',
+        'TSTypeAliasDeclaration',
+        'TSInterfaceDeclaration',
+        'TSAbstractClassDeclaration',
+        'TSModuleDeclaration',
+      ]
+      const exportedDecls = ast.body.filter(({ type, id, declarations }) => 
+        declTypes.includes(type) && 
+        (id && id.name === exportedName || declarations.find(d => d.id.name === exportedName))
```
Comment (julien1619) [inline src/ExportMap.js:546]: Hey! Here `declarations` is undefined unless `type === 'VariableDeclarations'` so it crashes the linter in some cases. I'll create another PR to fix it.

Comment (ljharb) [reply src/ExportMap.js:546]: Thanks!

**Comment by richard-lopes-ifood** [comment]
Nice fixing :rocket:! When is it going to be released?

**Pull request merged**